전남대 FE_서민지 - 2주차 과제

minji2219 opened the pull request:
안녕하세요! 멘토님 2주차 과제 step1 완료 후 1차 PR입니다.

### STEP1 질문
폴더 구조에 대한 질문입니다. main페이지에서만 사용하는 component같은 경우에는 page/main/해당컴포넌트 의 구조로 하는 것과 components/main/해당컴포넌트 이런식으로 components 하위 폴더에 위치 시켜야하는지 궁금합니다.

minji2219 committed:
Feat: react-router-dom 설치 및 router 구현

router 컴포넌트 생성
메인, 로그인, 내계정, 테마 페이지 생성 후 router로 각 path별 페이지 생성 및 연결

minji2219 committed:
Feat: 각 페이지 폴더로 관리

minji2219 committed:
Feat: Header 생성

minji2219 committed:
Feat: Footer 생성

minji2219 committed:
Feat: Layout 구현 및 router 설정

Header, Footer를 모든 페이지에서 볼 수 있게 Layout으로 생성
Outlet 이용

minji2219 committed:
Feat: Header 가운데 정렬

minji2219 committed:
Feat: 메인페이지 메인 멘트 컴포넌트 구현

minji2219 committed:
Feat: annual category 구현

기념일 마다의 선물 카테고리를 뜻하는 annual category 컴포넌트 구현

minji2219 committed:
Feat: Badge 컴포넌트 생성

minji2219 committed:
Feat: PersonCateogry 컴포넌트 추가

선물 랭킹 중 인물이 카테고리인 컴포넌트 생성

minji2219 committed:
Feat: SituationCategory 컴포넌트 생성

선물랭킹 중 상황을 선택하는 카테고리 컴포넌트 생성

minji2219 committed:
Rename: Theme로 파일명 변경

AnnualCateogry 컴포넌트 명을 Theme로 파일명 및 컴포넌트 명 변경

minji2219 committed:
Feat: Rankings 컴포넌트 생성

*22개의 더미 데이터 생성 후 사용
*실시간 급상승 선물랭킹 항목 보여주는 컴포넌트

minji2219 committed:
Feat: 불필요한 텍스트 삭제

minji2219 committed:
Feat: Footer 높이 재설정

minji2219 committed:
Feat: Header 상단에 sticky하게 설정

minji2219 committed:
Feat: cursor pointer 필요한 부분 스타일 추가

minji2219 committed:
Feat: theme dynamic route 설정

minji2219 committed:
Feat: a태그 css 리셋

minji2219 committed:
Feat: theme Header 섹션 구현

minji2219 committed:
Feat: theme페이지 goodslist 구현

minji2219 committed:
Feat: layout Header 스타일 수정

theme페이지 구현 중 header max-width로 인한 깨짐 발생으로 width수정

minji2219 committed:
Feat: login페이지 구현

minji2219 committed:
Feat: Header 로그인 버튼 로그인 페이지 link연결

minji2219 committed:
Feat: 로그인 페이지 layout 보여주기 제외

minji2219 committed:
Feat: myaccount페이지 구현

minji2219 committed:
Feat: a태그 focus될 때 outline 없애기

minji2219 committed:
Feat: Header 선물하기 로고 클릭시 메인페이지로 이동

minji2219 committed:
Feat: 불필요한 css 삭제

Link태그로 감싸 pointer css불필요해짐

minji2219 committed:
Feat: person category state로 관리

minji2219 committed:
Feat: situation category state로 관리

minji2219 committed:
Feat: key값 없음으로 인한 에러 수정

Theme 컴포넌트
Ranking컴포넌트
Situation category 컴포넌트

minji2219 committed:
Merge pull request #1 from minji2219/step1

2주차 Step1 구현

minji2219 committed:
Docs: step1과제 readme작성

minji2219 committed:
Feat: 경로 상수화

유지보수 등을 위해 경로 상수로 만들어 사용

minji2219 committed:
Docs: step2 해야할 일 readme작성

minji2219 committed:
Feat: UserInfo contextAPI 구현

username과 setusername을 value로 하는 contextAPI

minji2219 committed:
Feat: 로그인시 userName storage에 저장 및 context로 공유

userName sessionStorage에 저장
contextAPI로 userName 공유
로그인 후 전 페이지로 이동

minji2219 committed:
Feat: 로그인 여부에 따른 헤더 수정

로그인 되어 있을 경우 "내 계정"버튼
로그인 안되어 있을 경우 "로그인"버튼

minji2219 committed:
Feat: my-account는 로그인 된 사람만 접근 가능하게 구현

로그인 되지 않았을 때 my-account로 접근 시 login페이지로 이동

minji2219 committed:
Feat: my-account에서 로그아웃 구현

sessionStorage값 clear
context값 ''로 초기화

minji2219 committed:
Feat: 로그인, 내계정 상황에 따라 link수정

로그인 된 상태일 때는 내계정 클릭시 my-account로 이동
로그인 안된 상태일 때는 로그인 클릭시 login로 이동

minji2219 committed:
Feat: 로그아웃시 메인페이지로 redirect

minji2219 committed:
Feat: 초기 userName설정 변경

authToken의 값이 존재한다면 authToken값으로 초기화

minji2219 committed:
Merge pull request #2 from minji2219/step2

2주차 과제 step2구현

tlsehdfl commented on src/components/Layout/Header.tsx:
```diff
@@ -0,0 +1,42 @@
+          <Link to={ROUTE_PATHS.ROOT}>선물하기</Link>
+        </Logo>
+        <Link to={userName === auth ? '/my-account' : '/login'}>
+          {userName === auth ? '내 계정' : '로그인'}
```
`userName` 과 sessionStorage 의 authToken 을 비교해서 로그인 상태를 판단하는 이유는 뭘까요 ?

minji2219 replied:
꼼꼼한 리뷰 감사합니다!
userInfo context를 처음 만들 때 defaultUserInfo의 userName을 `''` 이런식으로 지정했다가
authToken값이 존재할 경우 해당 값으로 초기화 하는 것으로 코드를 바꿨습니다. 
처음처럼 코드를 작성했을 때 저 비교가 필요하다고 판단했는데 한 값만 넣어도 될거 같습니다!

혹시 저 상황에서는 auth값이 존재할 경우만 확인하면 될까요?

tlsehdfl commented on src/components/Routes/index.tsx:
```diff
@@ -0,0 +1,30 @@
+        {userName === auth ? (
+          <Route path={ROUTE_PATHS.MYPAGE} element={<MyAccount />} />
+        ) : (
+          <Route path={ROUTE_PATHS.MYPAGE} element={<Login />} />
```
로그인되지 않은 상태라면 path 자체도 Login page 에 경로여야 할 거 같습니다.

tlsehdfl commented on src/pages/Home/Badge.tsx:
```diff
@@ -0,0 +1,33 @@
+import { Button } from '@/components/common/Button';
+import { breakpoints } from '@/styles/variants';
+
+export const Badge = () => {
```
Badge 라는 네이밍이 해당 컴포넌트에 적합하지 않은것 같은 생각이 드네요. 더 나은 네이밍이 있을까요 ?

tlsehdfl commented on src/pages/Home/Badge.tsx:
```diff
@@ -0,0 +1,33 @@
+
+export const Badge = () => {
+  return (
+    <Wrapper>
```
원론적인 구분으로 `Container` 가 더 적합해 보입니다.
ref. https://velog.io/@na_0_i/Container%EC%99%80-Wrapper-%EB%91%90-%EC%9A%A9%EC%96%B4%EC%9D%98-%EC%B0%A8%EC%9D%B4%EC%A0%90%EC%9D%80-%EB%AC%B4%EC%97%87%EC%9D%BC%EA%B9%8C

tlsehdfl commented on src/pages/Home/Ranking/PersonCategory.tsx:
```diff
@@ -0,0 +1,77 @@
+  const [category, setCategory] = useState<string>('all');
+  const handleCategory = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setCategory(e.target.value);
+    console.log(category);
```
```suggestion
```

tlsehdfl commented on src/pages/Home/Ranking/PersonCategory.tsx:
```diff
@@ -0,0 +1,77 @@
+];
+
+export const PersonCategory = () => {
+  const [category, setCategory] = useState<string>('all');
```
`string` 타입이 아닌 category 에 해당하는 타입 지정이 필요합니다.

minji2219 replied:
type Category = {'all' | 'f' | 'm' | 's'}와 같이 type을 지정하라는 말씀이실까요?

tlsehdfl commented on src/pages/Home/Ranking/SituationCategory.tsx:
```diff
@@ -0,0 +1,72 @@
+];
+
+export const SituationCategory = () => {
+  const [category, setCategory] = useState('want');
```
이 부분 또한 `string` 타입이 아닌 `category` 에 해당하는 타입 지정이 필요합니다.

tlsehdfl commented on src/pages/Home/Ranking/SituationCategory.tsx:
```diff
@@ -0,0 +1,72 @@
+  const [category, setCategory] = useState('want');
+  const handleCategory = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setCategory(e.target.value);
+    console.log(category);
```
```suggestion
```

tlsehdfl commented on src/pages/Login/index.tsx:
```diff
@@ -0,0 +1,62 @@
+
+export const Login = () => {
+  const [name, setName] = useState<string>('');
+  const [pw, setPw] = useState<string>('');
```
```suggestion
  const [password, setPassword] = useState<string>('');
```

tlsehdfl commented on src/pages/Login/index.tsx:
```diff
@@ -0,0 +1,62 @@
+  const { setUserName } = useContext(UserInfo);
+
+  const hanldeClick = () => {
+    sessionStorage.setItem('authToken', name);
```
보통 authToken 이라고 하면 login api 호출하여 취득한 accessToken 을 말합니다.
유저의 이름을 sessionStorage 에 저장하여 authToken 처럼 사용하는건 지양해야 합니다.

tlsehdfl commented on src/pages/Login/index.tsx:
```diff
@@ -0,0 +1,62 @@
+  const hanldeClick = () => {
+    sessionStorage.setItem('authToken', name);
+    setUserName(name);
+    navigate(-1);
```
로그인 이후에는 로그인 된 상태로 홈으로 이동해야 합니다. navigation 뒤로가기를 하면 브라우저에서 앞으로 가기 했을 때 다시 로그인 페이지로 이동할거에요.

tlsehdfl commented on src/pages/Login/index.tsx:
```diff
@@ -0,0 +1,62 @@
+  width: 88px;
+`;
+const Container = styled.div`
+  // display: flex;
```
```suggestion
```

tlsehdfl commented on src/pages/Login/index.tsx:
```diff
@@ -0,0 +1,62 @@
+`;
+const Container = styled.div`
+  // display: flex;
+  // flex-direction: column;
```
```suggestion
```

tlsehdfl commented on src/pages/Login/index.tsx:
```diff
@@ -0,0 +1,62 @@
+const Container = styled.div`
+  // display: flex;
+  // flex-direction: column;
+  // gap: 40px;
```
```suggestion
```

tlsehdfl commented:
> 폴더 구조에 대한 질문입니다. main페이지에서만 사용하는 component같은 경우에는 page/main/해당컴포넌트 의 구조로 하는 것과 components/main/해당컴포넌트 이런식으로 components 하위 폴더에 위치 시켜야하는지 궁금합니다.

선택하기에 따라 다릅니다 ! 실무 입장에서는 보통 어플리케이션이 확장함을 염두하여 feature based structure 를 지향합니다.
말씀해주신 선택지의 전자에 해당합니다.

ref. https://maestros.io/structure-by-type-vs-feature

minji2219 committed:
Docs: step3 질문에 대한 답변 작성

minji2219 committed:
Merge pull request #3 from minji2219/step3

step3 질문에 대한 답변 작성

minji2219 committed:
Refactor: 피드백을 기반으로 코드 수정

*불필요한 context값 sessionStorage값 비교 제거
*불분명한 컴포넌트 네이밍 수정(Badge->SubMent)
*Wrapper, Container 구분
*불필요한 console.log()제거
*navigate 수정

minji2219 committed:
Refactor: 피드백을 기반으로 수정

*cateogry type string대신 해당 type 지정
*불필요한 console.log() 제거
*pw -> password로 변경
